feat: TextField can customize the Error, Label, and add elements before and after input (#3386)

GeorgeHulpoi opened the pull request:
## Description

https://github.com/payloadcms/payload/discussions/3386

- [x] I have read and understand the [CONTRIBUTING.md](https://github.com/payloadcms/payload/blob/main/CONTRIBUTING.md) document in this repository.

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Chore (non-breaking change which does not add functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Change to the [templates](../templates/) directory (does not affect core functionality)
- [ ] Change to the [examples](../examples/) directory (does not affect core functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes
- [x] I have made corresponding changes to the documentation

DanRibbens commented:
Hi @GeorgeHulpoi,
Thanks for the PR! We'd like to see the `BeforeInput` and `AfterInput` components changed to support an array. The reason for this is if a plugin needs to extend fields in a project or another plugin has already assigned these, they won't overwrite or have to wrap one another.

I would also like to see the docs change moved to https://payloadcms.com/docs/admin/components#fields instead of docs/fields/text and implement this change across all fields. This might not be applicable to the field `type: 'ui'`, but I think every field should have the same functionality ideally. 

The tests don't need to be changed for every field, as long as we're handling it the same in every place.

Are you able to do this?

DanRibbens reviewed (changes requested):
https://github.com/payloadcms/payload/pull/3747#issuecomment-1777458445

GeorgeHulpoi commented:
@DanRibbens there are the updated fields. Some of them don't have a `Label` or `Error`.  
Also, I added `BeforeInput`/`AfterInput` only for these fields that are basic and use `input`/`textarea` (where was possible). 

- [x] Text
- [x] Email
- [x] Textarea
- [x] Upload (without BeforeInput/AfterInput) 
- [x] Select (without BeforeInput/AfterInput)
- [ ] ~UI~
- [ ] ~RichText~
- [x] Relationship (without BeforeInput/AfterInput) 
- [x] RadioGroup (without BeforeInput/AfterInput)
- [x] Point
- [x] Number (BeforeInput/AfterInput only when `hasMany` is false)
- [x] JSON (without BeforeInput/AfterInput)
- [ ] ~Group~
- [x] Date (without BeforeInput/AfterInput)
- [ ] ~Collapsible~
- [x] Code (without BeforeInput/AfterInput)
- [ ] ~Blocks~
- [ ] ~Array~
- [x] Checkbox

Also, the request for docs it's still the same? Is not only the UI Field, there are fields that don't implement `BeforeInput`/`AfterInput`

DanRibbens commented:
I think the docs could be generic enough by saying "for fields with labels" in the table for Label as an example. People will figure out what is available through the typescript types and config validation.

Alternatively if you wanted to spell it all out on each docs field page you could. Whatever you think and we can always improve docs based on feedback.

GeorgeHulpoi commented:
@DanRibbens thanks for the answer. I moved and changed the doc for it. Please check twice because I'm not native in English and the texts may have mistakes.

DanRibbens commented on docs/admin/components.mdx:
```diff
@@ -426,6 +426,20 @@ All Payload fields support the ability to swap in your own React components. So,
+| Component   | Description                                        |
+| ----------- | -------------------------------------------------- |
+| **`Label`** | Override the default Label in the Field Component. |
+| **`Error`** | Override the default Label in the Field Component. |
```
We can list the new Component options available with the descriptions you have, but we should then link to new sections describing each.

For each new Component property allowed we will want to describe the props that these components take to better help people build their own custom components and possibly include simple example components like you have in your tests.

DanRibbens commented on docs/admin/components.mdx:
```diff
@@ -426,6 +426,20 @@ All Payload fields support the ability to swap in your own React components. So,
+
+| Property          | Description                                                                 |
+| ----------------- | --------------------------------------------------------------------------- |
+| **`BeforeInput`** | An array of elements that will be added before `input`/`textarea` elements. |
```
Again, we need to have this up in the list of possible components but then link to a section that describes them better. It probably could be `$$ BeforeInput and AfterInput` under one section that describes them.

DanRibbens commented on docs/admin/components.mdx:
```diff
@@ -426,6 +426,20 @@ All Payload fields support the ability to swap in your own React components. So,
 | **`Cell`**   | Used in the `List` view's table to represent a table-based preview of the data stored in the field. [More](#cell-component) |
 | **`Field`**  | Swap out the field itself within all `Edit` views. [More](#field-component)                                                 |
 
+There are fields that support **`Label`** and **`Error`** that are swappable likewise. The main idea of these properties is to extend existing fields without reimplementing them. The only fields that **DON'T** support are [Array](/docs/fields/array), [Blocks](/docs/fields/blocks), [Collapsible](/docs/fields/collapsible), [Group](/docs/fields/group), [Rich Text](/docs/fields/rich-text), [Row](/doc ...
```
I would replace this paragraph with the following, but keep your nicer formatting.

> As an alternative to replacing the entire Field component, you may want to keep the majority of the default Field component and only swap components within. This allows you to replace the Error or Label within a field component or add additional components inside the field with BeforeInput or AfterInput. BeforeInput and AfterInput is allowed in any  fields that don't contain other fields or RichText.

DanRibbens commented on docs/admin/components.mdx:
```diff
@@ -426,6 +426,20 @@ All Payload fields support the ability to swap in your own React components. So,
+| **`Label`** | Override the default Label in the Field Component. |
+| **`Error`** | Override the default Label in the Field Component. |
+
+Also, there are fields that support BeforeInput and AfterInput. Supported by [Checkbox](/docs/fields/checkbox), [Email](/docs/fields/email), [Number](/docs/fields/number), [Point](/docs/fields/point), [Text](/docs/fields/text), and [Textarea](/docs/fields/textarea) fields.
```
The paragraph I wrote it in the comment above replaces this one. Also you don't have Date field and I'm wondering why that is. I see you didn't implement it, what is your reason not to?

DanRibbens commented on packages/payload/src/admin/components/forms/field-types/DateTime/index.tsx:
```diff
@@ -60,9 +73,9 @@ const DateTime: React.FC<Props> = (props) => {
-      <Label htmlFor={path} label={label} required={required} />
+      <LabelComp htmlFor={path} label={label} required={required} />
       <div className={`${baseClass}__input-wrapper`} id={`field-${path.replace(/\./g, '__')}`}>
         <DatePicker
```
You don't think we could add render the `BeforeInput` or `AfterInput` components around the `<DatePicker>`?

GeorgeHulpoi replied:
Initially, I thought that the `<DatePicker/>` itself is a complex component and `BeforeInput` and `AfterInput` should be used only in fields that have `input` or `textarea`. But thinking now, it may be used to place tooltips.

DanRibbens reviewed (changes requested):
Looking very good overall!

Mostly interested in improving the docs at this point.

One other thing to look at before merging, are we exporting the props correctly to make this easy to implement? It is nice to not have to import from `/dist`.
